atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/transaction/impl/AdvisoryLockPreCommitCheckTest.java

```diff
@@ -0,0 +1,195 @@
+    }
+
+    @Test
+    public void checkFailsIfOnlySomeLegacyLocksAreRefreshed() {
+        LockService lockService = mock(LockService.class);
+        when(lockService.refreshLockRefreshTokens(any())).thenReturn(ImmutableSet.of(TOKEN_1));
+
+        assertThatThrownBy(AdvisoryLockPreCommitCheck.forLockServiceLocks(TOKENS, lockService)::throwIfLocksExpired)
+                .isInstanceOf(TransactionLockTimeoutException.class);
+    }
+
+    @Test
+    public void checkFailsIfNoLegacyLocksAreRefreshed() {
+        LockService sadLockService = mock(LockService.class);
+        when(sadLockService.refreshLockRefreshTokens(any())).thenReturn(ImmutableSet.of());
+
+        assertThatThrownBy(AdvisoryLockPreCommitCheck.forLockServiceLocks(TOKENS, sadLockService)::throwIfLocksExpired)
+                .isInstanceOf(TransactionLockTimeoutException.class);
+    }
+
+    @Test
+    public void checkDelegatesAndFailsOnlyWhenChecked() {
+        LockService sadLockService = mock(LockService.class);
+        when(sadLockService.refreshLockRefreshTokens(any())).thenReturn(ImmutableSet.of());
+
+        AdvisoryLockPreCommitCheck preCommitCheck = AdvisoryLockPreCommitCheck.forLockServiceLocks(
+                TOKENS, sadLockService);
+
+        verify(sadLockService, never()).refreshLockRefreshTokens(any());
+        try {
+            preCommitCheck.throwIfLocksExpired();
+        } catch (Exception e) {
+            // expected
+        }
+        verify(sadLockService).refreshLockRefreshTokens(any());
+    }
+
+    @Test
+    public void propagatesExceptionsIfLockServiceThrows() {
+        IllegalStateException exception = new IllegalStateException("something bad");
+        LockService lockService = mock(LockService.class);
+        when(lockService.refreshLockRefreshTokens(any())).thenThrow(exception);
+
+        assertThatThrownBy(AdvisoryLockPreCommitCheck.forLockServiceLocks(TOKENS, lockService)::throwIfLocksExpired)
+                .isInstanceOf(IllegalStateException.class)
+                .isNotInstanceOf(TransactionLockTimeoutException.class);
+    }
+
+    @Test
+    public void doesNotDelegateCheckIfNoLegacyLockTokensAreSupplied() {
+        LockService lockService = mock(LockService.class);
+        AdvisoryLockPreCommitCheck.forLockServiceLocks(ImmutableList.of(), lockService).throwIfLocksExpired();
+        verifyNoMoreInteractions(lockService);
+    }
+
+    @Test
+    @SuppressWarnings("unchecked") // we know throughout that we refer to Sets of LockRefreshToken
+    public void checkPassesIfAsyncLocksAreRefreshed() {
+        TimelockService timelockService = mock(TimelockService.class);
+        when(timelockService.refreshLockLeases(any())).thenAnswer(invocation -> invocation.getArguments()[0]);
+
+        AdvisoryLockPreCommitCheck.forAsyncLockServiceLocks(TOKENS_V2, timelockService).throwIfLocksExpired();
+
+        // We need to check that the element contains-exactly-in-any-order the same elements
```

Comment:
Yep. I dug a bit harder and it turns out that this can actually be simplified quite a lot, to something like:

```java
    private void verifyCalledWithLegacyTokens(LockService lockService) {
        // Note: eq is too strong, because it requires that the collection type matches the type of TOKENS
        verify(lockService).refreshLockRefreshTokens(
                (Set<LockRefreshToken>) argThat(containsInAnyOrder(TOKENS.toArray())));
    }
```